R/verify.R

```diff
@@ -0,0 +1,83 @@
+#' Check Code Reproducibility
+#'
+#' Checks whether code in `teal_data` reproduces the stored objects.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' verify(tdata3)
+#' }
+#' @name verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#' @rdname verify
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
+  if (x@verified) {
+    return(x)
+  }
+  new_teal_data <- eval_code(teal_data(), x@code)
+
+  if (inherits(new_teal_data, "qenv.error")) {
+    stop(conditionMessage(new_teal_data), call. = FALSE)
+  }
+
+  reproduced <- isTRUE(all.equal(x@env, new_teal_data@env))
+  if (reproduced) {
+    x@verified <- TRUE
+    methods::validObject(x)
+    x
+  } else {
+    names_diff <- setdiff(names(x@env), names(new_teal_data@env))
+
+    objects_diff <- vapply(
+      intersect(names(x@env), names(new_teal_data@env)),
+      function(element) {
+        isTRUE(all.equal(x@env[[element]], new_teal_data@env[[element]]))
+      },
+      logical(1)
+    )
+
+    names_diff_other <- setdiff(names(new_teal_data@env), names(x@env))
+
+    error <- c(
+      "\nCode verification failed at object(s):\n",
+      paste(c(names_diff, names(which(!objects_diff))), collapse = "\n")
+    )
+
+    if (length(names_diff_other)) {
+      error <- c(
+        error,
+        "\nObject(s) created with code that do not exist in teal_data:\n",
+        paste(names_diff_other, collapse = "\n")
+      )
+    }
```

Comment:
Small _(optional)_ suggestion to build the error message. 

Given that `\n` is as a separator in all lines.

```suggestion
    error <- c(
      "Code verification failed at object(s):",
      names_diff,
      names(which(!objects_diff))
    )

    if (length(names_diff_other)) {
      error <- c(
        error,
        "Object(s) created with code that do not exist in teal_data:",
        names_diff_other
      )
    }

    stop(paste("", error, collapse = "\n"))
```

Reply:
Or this one, that adds a "  * " before every "name"

```suggestion
    error <- c(
      "Code verification failed at object(s):",
      paste0("  * ", c(names_diff, names(which(!objects_diff))))
    )
    if (length(names_diff_other)) {
      error <- c(
        error,
        "Object(s) created with code that do not exist in teal_data:",
        paste0("  * ", names_diff_other)
      )
    }
```

<details>

<summary>See example</summary>

``` r
names_diff <- "x2"
objects_diff <- c(x1 = FALSE)
names_diff_other <- "xOther"

    error <- c(
      "Code verification failed at object(s):",
      paste0("  * ", c(names_diff, names(which(!objects_diff))))
    )
    if (length(names_diff_other)) {
      error <- c(
        error,
        "Object(s) created with code that do not exist in teal_data:",
        paste0("  * ", names_diff_other)
      )
    }
paste(error, collapse = "\n") |> cat()
#> Code verification failed at object(s):
#>   * x2
#>   * x1
#> Object(s) created with code that do not exist in teal_data:
#>   * xOther
```

</details>

Reply:
I like the dot, as in a bullet list! Thanks for the suggestion. I will also use your simplified way of pasting/gathering the error message as it's way cleaner! Thanks

Reply:
Not bad but I'll go you one further: take the `paste` out of condition functions. Will save you some headache when submitting to Bioconductor.

Reply:
ins't teal.data planned to be submitted on CRAN not Bioconductor?